### PR TITLE
Use DisposableElementsAttr for ZHigh constant propagation

### PR DESCRIPTION
**Pull request by tungld**
- This patch reverts the PR #2917 
   - The old way in #2917 is to propose a ConstantOpInterface and delay the stickification until storing constants-to-file. Two main disadvantages are: 1) constant propagation for ztensor at zhigh dialect is off, 2) it does not work if `store-constants-to-file` is off.
- This patch extends the constant prop mechanism (that currently applies to `ONNXConstantOp`) to `ZHighStickifiedConstantOp`, so that the zhigh constant prop can benefit from the memory management for `DisposableElementsAttr`: In particular, it does:
  - Make all ZHigh optimizations/rewritings work with` DisposableElementsAttr`
  - Change `ZHighStikifiedConstantOp` to accept `DisposableElementsAttr`. Its parser and printer are changed to read/write from/to `DenseElementsAttr` for lit tests.
  - Change `ZHighConstantPropagationPass` so that it reads data directly from `DisposableElementsAttr` instead of `DenseElementsAttr`.
  - Add two passes: `ZHighDisposableGabageCollector` and `ZHighScrubDisposablePass` to manage buffers used by `ZHighStickifiedConstantOp`

 
Quick experiment: the peak compile memory consumptions of #2917 and this PR when compiling the gpt2-large model for NNPA (744M parameters, the constant file's size is 3.2GB) are quite similar, both are about 9GB.

This patch contains the reverting code so it's no easy to follow. To ease the review, I merge all new changes (not the reverting code) into a single commit: https://github.com/onnx/onnx-mlir/pull/3013/commits/265ff9029c151f7a4a9473f23015414cd201f7e2. Please look at this commit for review.

### CODE REVIEW

**Comment by tungld** [comment]
> You are using Soren's approach of applying "logical" operations to the constants

Yes, I extend it for ZHigh operations so the same approach is used for both ONNX and ZHigh until lowering to krnl. We can extend it to cover krnl operations but it needs more work and I didn't do it in this PR.

**Review by imaihal — APPROVED** [review]
LGTM!

File: src/Accelerators/NNPA/Transform/ZHigh/ZHighConstPropagation.cpp [hunk 1]
```diff
@@ -237,10 +287,111 @@ namespace {
...
+      return failure();
+    }
+
+    // Rewrite
+    Value stickifiedVal =
+        createConstantForStick(rewriter, output, input, layout);
+    replaceOpAndGC(rewriter, stickOp, stickifiedVal);
+    return success();
+  }
+};
```
[the remark this answers is not in the document]
Comment (tungld) [reply src/Accelerators/NNPA/Transform/ZHigh/ZHighConstPropagation.cpp:336]: Yes, I don't know how to do that with tablegen, let me know if you know how to do it. Thanks!

**Comment by AlexandreEichenberger** [comment]
Can you post here for ref the improvements you got, just for future reference purpose. Does not need to be super detailed. Thanks

**Comment by imaihal** [comment]
> Can you post here for ref the improvements you got, just for future reference purpose. Does not need to be super detailed. Thanks

I put the measurement results of gpt2-large and Mistral-7b.
In gpt2-large, the peak memory usage reduced from 8.9 GB to 7.4 GB, and compilation time becomes faster from 5 min 22sec to 4 min 30 sec.

image

In gpt2-large, the peak memory usage reduced from 33.2 GB to 27.9 GB, and compilation time becomes faster from 17 min 4 sec to 13 min 58 sec.

image

**Pull request merged**